### PR TITLE
Add gpu/non-gpu resource sets for multi_task support

### PR DESCRIPTION
**Pull request by shuds13**
Note: When merge, update [Optimas branch](https://github.com/optimas-org/optimas/pull/88/files#diff-4d7c51b1efe9043e44439a949dfd92e5827321b34082903477fd04876edb7552R2)

Adding gpu/non-gpu resource sets for multi_task support.

Contains a new regression test `test_GPU_variable_resources_multi_task.py`

and a functionality test `test_mpi_gpu_settings_mock_nodes_multi_task.py`

When there are more resource sets on a node than GPUs, some resource sets will be GPU resource sets and some non-GPU resource sets.

The generator specifies, alongside `resource_sets`, a  boolean `use_gpus` as True or False. Both are `gen_specs["out"]`.
Alternatively, the generator can specify `num_procs` and `num_gpus` gen_specs['out'] options

### CODE REVIEW

File: docs/resource_manager/overview.rst [hunk 1]
```diff
@@ -14,18 +14,30 @@ these supplied resources, and if not given any of ``num_nodes``, ``num_procs``,
...
 Variable resource assignment
 ----------------------------
 
+.. note::
+    As of version 0.10.0, the concept of resource sets is not needed. The generator
+    can use special ``gen_specs["out"]`` fields of ``num_procs`` and ``num_gpus`` for each
+    simulation generated. These will be used to assign resources and will be
+    automatically passed through and used by the executor (if no other run configuration
+    is given in the submit line). Furthermore, GPUs will be automatically assigned
+    in the correct way for the given system (including nVIDA, AMD and Intel GPUs);
```
Comment (jlnav) [inline docs/resource_manager/overview.rst:29]: "Nvidia"

File: docs/resource_manager/overview.rst [hunk 2]
```diff
@@ -14,18 +14,30 @@ these supplied resources, and if not given any of ``num_nodes``, ``num_procs``,
 or ``procs_per_node`` in the submit function, it will try to use all nodes and
 CPU cores available to the worker.
 
-Detected resources can be overridden using the libE_specs option :ref:`resource_info<resource_info>`.
+Detected resources can be overridden using the libE_specs option
```
Comment (jlnav) [inline docs/resource_manager/overview.rst:17]: monospace, ``libE_specs``

File: docs/resource_manager/overview.rst [hunk 3]
```diff
@@ -14,18 +14,30 @@ these supplied resources, and if not given any of ``num_nodes``, ``num_procs``,
...
 ----------------------------
 
+.. note::
+    As of version 0.10.0, the concept of resource sets is not needed. The generator
+    can use special ``gen_specs["out"]`` fields of ``num_procs`` and ``num_gpus`` for each
+    simulation generated. These will be used to assign resources and will be
+    automatically passed through and used by the executor (if no other run configuration
+    is given in the submit line). Furthermore, GPUs will be automatically assigned
+    in the correct way for the given system (including nVIDA, AMD and Intel GPUs);
+    you do not need to set CUDA_VISIBLE_DEVICES or equivalent. Example:
```
Comment (jlnav) [inline docs/resource_manager/overview.rst:30]: ``CUDA_VISIBLE_DEVICES``

File: libensemble/executors/mpi_runner.py [hunk 4]
```diff
@@ -196,6 +207,29 @@ def _assign_gpus(self, task, resources, nprocs, nnodes, ppn, ngpus, extra_args,
 
         return nprocs, nnodes, ppn, extra_args
 
+    def _adjust_procs(self, nprocs, ppn, nnodes, ngpus, resources):
```
Comment (jlnav) [inline libensemble/executors/mpi_runner.py:210]: This *looks* refactor-able, but I may be wrong

File: libensemble/gen_funcs/persistent_sampling.py [hunk 5]
```diff
@@ -25,10 +35,8 @@ def persistent_uniform(_, persis_info, gen_specs, libE_info):
         `test_persistent_sampling.py <https://github.com/Libensemble/libensemble/blob/develop/libensemble/tests/regression_tests/test_persistent_sampling.py>`_ # noqa
         `test_persistent_sampling_async.py <https://github.com/Libensemble/libensemble/blob/develop/libensemble/tests/regression_tests/test_persistent_sampling_async.py>`_ # noqa
     """
-    ub = gen_specs["user"]["ub"]
-    lb = gen_specs["user"]["lb"]
-    n = len(lb)
-    b = gen_specs["user"]["initial_batch_size"]
+
+    b, n, lb, ub = _get_user_params(gen_specs["user"])
```
Comment (jlnav) [inline libensemble/gen_funcs/persistent_sampling.py:39]: I like this, maybe we ought to encourage this sort of extraction in the future (or extract user params into *args?)

File: libensemble/tools/alloc_support.py [hunk 6]
```diff
@@ -4,7 +4,11 @@
 
 from libensemble.message_numbers import EVAL_GEN_TAG, EVAL_SIM_TAG
 from libensemble.resources.resources import Resources
-from libensemble.resources.scheduler import InsufficientFreeResources, ResourceScheduler  # noqa: F401
+from libensemble.resources.scheduler import (
+    InsufficientFreeResources,
+    InsufficientResourcesError,
```
Comment (jlnav) [inline libensemble/tools/alloc_support.py:9]: What's the difference between each Insufficient exception?

**Review by jlnav — APPROVED** [review]

**Review by jmlarson1 — APPROVED** [review]

File: libensemble/gen_funcs/persistent_sampling.py (hunk 5 above)
Comment (jmlarson1) [reply libensemble/gen_funcs/persistent_sampling.py:39]: I agree, but couldn't that introduce namespace conflicts if everything is automatically extracted?

**Pull request merged**